src/vm/types/signatures.rs

```diff
@@ -132,367 +72,67 @@ impl FunctionType {
+            } else {
+                Ok(list_data)
             }
         }
     }
-}
 
-impl PrincipalData {
-    pub fn parse_qualified_contract_principal(literal: &str) -> Result<PrincipalData> {
-        let split: Vec<_> = literal.splitn(2, ".").collect();
-        if split.len() != 2 {
-            return Err(RuntimeErrorType::ParseError(
-                "Invalid principal literal: expected a `.` in a qualified contract name".to_string()).into());
-        }
-        let sender = Self::parse_standard_principal(split[0])?;
-        let name = split[1].to_string();
-        
-        Ok(PrincipalData::QualifiedContractPrincipal { sender, name })
-    }
-
-    pub fn parse_standard_principal(literal: &str) -> Result<StandardPrincipalData> {
-        let (version, data) = c32::c32_address_decode(&literal)
-            .map_err(|x| { RuntimeErrorType::ParseError(format!("Invalid principal literal: {}", x)) })?;
-        if data.len() != 20 {
-            return Err(RuntimeErrorType::ParseError(
-                "Invalid principal literal: Expected 20 data bytes.".to_string()).into());
+    pub fn get_list_item_type(&self) -> TypeSignature {
+        if self.dimension == 0 {
+            panic!("Invalid dimension of 0")
+        } else if self.dimension == 1 {
+            TypeSignature::Atom(self.atomic_type.clone())
+        } else {
+            let list_type = ListTypeData { 
+                atomic_type: self.atomic_type.clone(),
+                max_len: self.max_len.clone(),
+                dimension: self.dimension - 1 };
+            TypeSignature::List(list_type)
         }
-        let mut fixed_data = [0; 20];
-        fixed_data.copy_from_slice(&data[..20]);
-        Ok(StandardPrincipalData(version, fixed_data))
-    }
-
-    pub fn deserialize(json: &str) -> PrincipalData {
-        serde_json::from_str(json)
-            .expect("Failed to deserialize vm.PrincipalData")
-    }
-    pub fn serialize(&self) -> String {
-        serde_json::to_string(self)
-            .expect("Failed to serialize vm.PrincipalData")
     }
-}
 
-impl fmt::Display for PrincipalData {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match self {
-            PrincipalData::StandardPrincipal(sender) => {
-                let c32_str = c32::c32_address(sender.0, &sender.1[..])
-                    .unwrap_or_else(|_| "INVALID_C32_ADD".to_string());
-                write!(f, "'{}", c32_str)                
-            },
-            PrincipalData::ContractPrincipal(contract_name) => {
-                write!(f, "'CT{}", contract_name)
-            },
-            PrincipalData::QualifiedContractPrincipal { sender, name } => {
-                let c32_str = c32::c32_address(sender.0, &sender.1[..])
-                    .unwrap_or_else(|_| "INVALID_C32_ADD".to_string());
-                write!(f, "'CT{}.{}", c32_str, name)
-            }
+    pub fn size(&self) -> Result<i128> {
```

Comment:
Do you think we could structure lists so that the size calculation simply can't overflow?  We'd need to make it so you can't construct a list that has the capacity for so many items that an overflow in this method is possible (which is probably desirable just for resource utilization purposes).

Reply:
Yep -- I think this will be more or less addressed in the value bounding branch. It will still perform checked arithmetic for reasons mentioned elsewhere, but in non-construction contexts, I think it can be made infallible.